### PR TITLE
CVSL-138 Adding change type 

### PR DESCRIPTION
**Pull request by andrewrlee**
Also starting to return info about conditions that have been removed in later version of the policy

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctionsTest.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        code = "code-1", text = "some-text", data = listOf(
+          AdditionalConditionData(field = "ph-1")
+        )
+      )
+      assertThat(licencePolicyChanges(listOf(condition), listOf(policyCondition), placeHolders)).isEmpty()
+    }
+
+    @Disabled("Should pass?")
+    @Test
+    fun `if new fields added in policy then there should be a change?`() {
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctionsTest.kt:56]: Is this a bug?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctionsTest.kt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+          AdditionalConditionData(field = "ph-1")
+        )
+      )
+      val policyCondition = policyCondition().copy(code = "code-1", text = "some-text")
+
+      assertThat(licencePolicyChanges(listOf(condition), listOf(policyCondition), placeHolders)).isEmpty()
+    }
+
+    @Disabled("Should pass?")
+    @Test
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctionsTest.kt:58]: is this a bug? 

when the stored condition has no placeholders but the new policy version has some, it should return it as a change?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicencePolicyService.kt [hunk 3]
```diff
@@ -17,11 +17,16 @@ data class ConditionChangesByType<T, U>(
...
+  DELETED, REPLACED, NEW_OPTIONS, TEXT_CHANGE
+}
+
 data class LicenceConditionChanges(
+  val changeType: ConditionChangeType,
   val code: String,
   val sequence: Int?,
   val previousText: String,
-  val currentText: String,
+  val currentText: String?,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicencePolicyService.kt:29]: current text will be absent when the condition is deleted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt [hunk 4]
```diff
@@ -53,14 +57,26 @@ fun <T : ILicenceCondition> getPolicyPlaceholders(conditions: List<T>): HashMap<
...
+    val textHasChanged = match.text != condition.text
+    val dataChanges = condition.data.filter { d -> policyPlaceholders[condition.code]?.containsValue(d.field) == false }
+    val dataHasChanged = dataChanges.isNotEmpty()
+    when {
+      textHasChanged -> TEXT_CHANGE.of(condition, match.text, dataChanges)
+      dataHasChanged -> NEW_OPTIONS.of(condition, match.text, dataChanges)
+      else -> null
+    }
+  } else {
+    DELETED.of(condition, null, emptyList())
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt:73]: need to distinguish between REPLACED and DELETED here

**Comment by andrewrlee** [comment]
question around, why are the conditions that seem like they need to be deleted still present in the v2.1 config? 
e.g: [this](https://github.com/ministryofjustice/create-and-vary-a-licence/pull/387/files#diff-d9f6b308443c5b9c14f161b5c0fa02572ffcd4692260c287765ab5517ffc5941R20)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt [hunk 5]
```diff
@@ -53,14 +57,26 @@ fun <T : ILicenceCondition> getPolicyPlaceholders(conditions: List<T>): HashMap<
...
+    val textHasChanged = match.text != condition.text
+    val dataChanges = condition.data.filter { d -> policyPlaceholders[condition.code]?.containsValue(d.field) == false }
+    val dataHasChanged = dataChanges.isNotEmpty()
+    when {
+      textHasChanged -> TEXT_CHANGE.of(condition, match.text, dataChanges)
+      dataHasChanged -> NEW_OPTIONS.of(condition, match.text, dataChanges)
+      else -> null
+    }
+  } else {
+    DELETED.of(condition)
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt:73]: still need to distinguish between REPLACED and DELETED here
probably help to sort based on sequence too.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctionsTest.kt (hunk 2 above)
Comment (SebNegahbanMoJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctionsTest.kt:61]: For visibility:
We have discussed this on a call and the outcome was that we believe that this is correct expected behaviour, but will need to confirm. As there don't seem to be any changes of this type in the current policy changes (and with the imminent deadline), we're going to progress with the current behaviour and consider this post-release.

**Review by SebNegahbanMoJ — COMMENTED** [review]
Approved with the caveat that I'm by no means fluent in Kotlin. The logic seems reasonable, but I'd ask for eyes from @mgreen-mojhub before giving it a green tick

**Pull request closed without merging**